kahypar/partition/bin_packing/bin_packing_algorithms.h

```diff
@@ -0,0 +1,171 @@
+ *
+ ******************************************************************************/
+
+#pragma once
+
+#include <vector>
+
+#include "kahypar/definitions.h"
+#include "kahypar/datastructure/binary_heap.h"
+
+namespace kahypar {
+namespace bin_packing {
+using kahypar::ds::BinaryMinHeap;
+
+/*
+* An (online) bin packing algorithm is represented by a class that holds the internal state
+* and provides methods to insert elements or add initial weight to a bin. Additionally, it
+* is supported to lock a bin, which means that no more elements can be inserted to this bin.
+* For a type BPAlgorithm, the following operations must be available:
+* 1)
+*   PartitionID num_bins = ...;
+*   HypernodeWeight max_bin_weight = ...;
+*   BPAlgorithm alg(num_bins, max_bin_weight);
+* 2)
+*   PartitionID bin = ...;
+*   HypernodeWeight weight = ...;
+*   alg.addWeight(bin, weight);
+* 3)
+*   HypernodeWeight weight = ...;
+*   PartitionID resulting_bin = alg.insertElement(weight);
+* 4)
+*   ParititionID bin = ...;
+*   alg.lockBin(bin);
+* 5)
+*   ParititionID bin = ...;
+*   HypernodeWeight weight = alg.binWeight(bin);
+* 6)
+*   PartitionID numBins = alg.numBins();
+*/
+
+// Worst Fit algorithm - inserts an element to the bin with the lowest weight.
+class WorstFit {
+  public:
+    WorstFit(const PartitionID num_bins, const HypernodeWeight /*max*/) :
+      _bin_queue(num_bins),
+      _weights(),
+      _num_bins(num_bins) {
+      for (PartitionID i = 0; i < num_bins; ++i) {
+        _bin_queue.push(i, 0);
+      }
+    }
+
+    void addWeight(const PartitionID bin, const HypernodeWeight weight) {
+      ASSERT(bin >= 0 && bin < _num_bins, "Invalid bin id: " << V(bin));
+
+      _bin_queue.increaseKeyBy(bin, weight);
+    }
+
+    PartitionID insertElement(const HypernodeWeight weight) {
+      ASSERT(weight >= 0, "Negative weight.");
+      ASSERT(!_bin_queue.empty(), "All available bins are locked.");
+
+      // assign node to bin with lowest weight
+      PartitionID bin = _bin_queue.top();
+      _bin_queue.increaseKeyBy(bin, weight);
+      return bin;
+    }
+
+    void lockBin(const PartitionID bin) {
+      ASSERT(bin >= 0 && bin < _num_bins, "Invalid bin id: " << V(bin));
+      ASSERT(_bin_queue.contains(bin), "Bin already locked.");
+
+      if (_weights.empty()) {
+        _weights.resize(_num_bins, 0);
+      }
+      _weights[bin] = _bin_queue.getKey(bin);
+      _bin_queue.remove(bin);
+    }
+
+    HypernodeWeight binWeight(const PartitionID bin) const {
+      ASSERT(bin >= 0 && bin < _num_bins, "Invalid bin id: " << V(bin));
+
+      return _bin_queue.contains(bin) ? _bin_queue.getKey(bin) : _weights[bin];
+    }
+
+    PartitionID numBins() const {
+      return _num_bins;
+    }
+
+  private:
+    BinaryMinHeap<PartitionID, HypernodeWeight> _bin_queue;
+    std::vector<HypernodeWeight> _weights;
+    PartitionID _num_bins;
+};
+
+// First Fit algorithm - inserts an element to the first fitting bin.
+class FirstFit {
```

Comment:
Do we still need both implementations or do we only use one? If so, it might make sense to remove the unused implementation to reduce unused code.

Reply:
Well, it depends. Worst fit is the default, but in my thesis there were also a few cases where first fit performed better.

(It can be selected with the i-bp-algorithm parameter.)